File: apina-core/src/main/java/fi/evident/apina/java/reader/ClassPathScanner.java
package fi.evident.apina.java.reader;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.jar.JarInputStream;
import java.util.zip.ZipEntry;

import static java.nio.file.Files.isRegularFile;

/**
 * Reads class metadata for classes in classpath.
 */
final class ClassPathScanner {

    private static final Logger log = LoggerFactory.getLogger(ClassPathScanner.class);

    /**
     * Executes given processor for all classes
     */
    public static void processAllClasses(Classpath classpath, ClassPathResourceProcessor processor) throws IOException {
        for (Path classpathEntry : classpath.getRoots())
            processAllClasses(classpathEntry, processor);
    }

    private static void processAllClasses(Path classpathEntry, ClassPathResourceProcessor processor) throws IOException {
        log.debug("Scanning for classes in {}", classpathEntry);

        if (Files.isDirectory(classpathEntry)) {
            forEachClassUnderDirectory(classpathEntry, processor);

        } else if (isRegularFile(classpathEntry) && classpathEntry.toString().endsWith(".jar")) {
            forEachClassInArchive(classpathEntry, processor);

        } else if (Files.notExists(classpathEntry)) {
            log.warn("Skipping nonexistent classpath entry: {}", classpathEntry);

        } else {
            throw new IOException("invalid classpath entry: " + classpathEntry);
        }
    }

    private static void forEachClassInArchive(Path archivePath, ClassPathResourceProcessor processor) throws IOException {
        try (JarInputStream jar = new JarInputStream(Files.newInputStream(archivePath))) {
            while (true) {
                ZipEntry entry = jar.getNextEntry();
                if (entry == null)
                    break;

                if (entry.getName().endsWith(".class")) {
                    log.trace("Processing class-file {} from {}", entry.getName(), archivePath);
                    processor.process(jar);
                }
            }
        }
    }

    private static void forEachClassUnderDirectory(Path directory, ClassPathResourceProcessor processor) throws IOException {
        try {
            Files.walk(directory)
                    .filter(p -> isRegularFile(p) && p.toString().endsWith(".class"))
                    .forEach(file -> {
                        log.trace("Processing class-file {}", file);
                        try (InputStream in = new BufferedInputStream(Files.newInputStream(file))) {
                            processor.process(in);
                        } catch (IOException e) {
                            throw new UncheckedIOException(e);
                        }
                    });
        } catch (UncheckedIOException e) {
            throw e.getCause();
        }
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/java/reader/ClassPathScanner.kt
package fi.evident.apina.java.reader

import fi.evident.apina.utils.SkipZipPrefixStream
import org.slf4j.LoggerFactory
import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Files.isRegularFile
import java.nio.file.Path
import java.util.jar.JarInputStream

private object ClassPathScanner

private val log = LoggerFactory.getLogger(ClassPathScanner::class.java)

/**
 * Executes given processor for all classes
 */
fun Classpath.processAllClasses(processor: (InputStream) -> Unit) {
    for (classpathEntry in roots)
        classpathEntry.processAllClasses(processor)
}

private fun Path.processAllClasses(processor: (InputStream) -> Unit) {
    log.debug("Scanning for classes in {}", this)

    when {
        Files.isDirectory(this) ->
            forEachClassUnderDirectory(processor)

        isRegularFile(this) && hasExtension(".jar") ->
            forEachClassInArchive(processor)

        Files.notExists(this) ->
            log.warn("Skipping nonexistent classpath entry: {}", this)

        else ->
            throw IOException("invalid classpath entry: $this")
    }
}

private fun Path.forEachClassInArchive(processor: (InputStream) -> Unit) {
    JarInputStream(SkipZipPrefixStream(Files.newInputStream(this))).use { jar ->
        while (true) {
            val entry = jar.nextEntry ?: break

            if (entry.name.endsWith(".class")) {
                log.trace("Processing class-file {} from {}", entry.name, this)
                processor(jar)
            }
        }
    }
}

private fun Path.forEachClassUnderDirectory(processor: (InputStream) -> Unit) {
    Files.walk(this)
            .filter { p -> isRegularFile(p) && p.toString().endsWith(".class") }
            .forEach { file ->
                log.trace("Processing class-file {}", file)

                BufferedInputStream(Files.newInputStream(file)).use(processor)
            }
}

private fun Path.hasExtension(vararg extensions: String): Boolean {
    val str = toString()
    return extensions.any { str.endsWith(it) }
}
